### DISCUSSION TITLE
(done) Roadmap to 4.0

### DISCUSSION THREAD

**Discussion by ghiscoding** [ghiscoding/slickgrid-universal#1108] [Ideas]
### TODOs
- [x] create [Migration Guide 4.x](https://github.com/ghiscoding/slickgrid-universal/wiki/Migration-to-4.x)
- [x] use newer SlickGrid ESM builds (ref PR #1070)
  - [x] after PR #1070 is merged, it might be a good idea to actually drop SlickGrid dependency entirely and copy/paste the newer SlickGrid ESM directly into Slickgrid-Universal to drop external deps
     - [x] wait for official SlickGrid 5.0 to be in users hand for about a month to see if any bugs showed up
     - [x] Move SlickGrid core files into Slickgrid-Universal (ref PR #1135)
     - [x] remove unused jQuery animation wrappers
     - [x] Remove duplicate methods like "createDomElement()", "calculateAvailableSpace()", "emptyElement()",
     - [x] remove unused `AutoSize` and also all interface props (ie: in `Column`), also review all `slick.core` enums that we no longer need, might not need `gridAutosizeColsMode` grid option (commits: [af513e27](https://github.com/6pac/SlickGrid/commit/af513e27a5a984d7aa5bfe79e0eb9508481f6bf8), [9fc3ab9d](https://github.com/6pac/SlickGrid/commit/9fc3ab9db0bbc47d37cafe4e278bd82feaf45b65))
        - that actually removed 4Kb from final build size :)
     - [x] migrating SlickGrid into the project will decrease unit test coverage since no unit tests exists for SlickGrid, SlickDataView & SlickCore, so we need to add some tests but that could be done few months after the release
        - after merging it was around 22% decrease but I managed to add more tests and decreased it to 14% (to be improved in the future)
     - [x] review Cypress tests from SlickGrid and see if we missed any in our project (i.e. "Example 1 Basic Grid" has tests that we don't have)
     - [x] replace all Formatters that uses html string with native html, compare build size since it might impact its size
     - [x] replace all GroupTotalsFormatters that uses html string with native html
     - [x] possibly drop `AlignRight`, `Center`, `Bold`, `Italic`, `Uppercase`, ... Formatters since they could & should be replaced with simple `cssClass` (or `headerCssClass` for column titles)
        - [x] add `Formatters.multiple` console error when Formatter doesn't exist
     - [x] `slick.groupitemmetadataprovider` has an html string formatter that should be converted to native element
     - [x] calling `applyHtmlCode()` with an empty string shouldn't need to call `innerHTML` at all
     - [x] `removeHtmlTags` is calling `innerHTML`, we need to find a better alternative
        - we could use 3rd party lib [striptags](https://github.com/ericnorris/striptags)
     - [x] some Formatters and other sections use inline `style` but that is not CSP compliant
     - [x] current Formatter result to check object could be incorrect when it returns a Proxy (Salesforce) which is an object and tries to get FormatterObject `html` or `text`
     - [x] see if we can use more `DocumentFragment` in Formatters instead of returning strings
        - see PR #1215 
     - [x] rename `sanitizerHtmlOptions` to `sanitizerOptions`
     - [x] remove any IE related code, should we keep `fastSort()`?
     - [x] investigate if we can remove default import Rollup patches (Moment, SortableJS, DOMPurify, Autocompleter)
       - [x] might need to switch imports from `import * as autocompleter_ from 'autocompleter'` to `import autocompleter from 'autocompleter'`
     - [x] move `slick.core` Utils and consolidate them with Universal's utils (either DomUtils or Utils in Services folder)
     - [x] possibly rename some DomUtils methods to shorter name, like `getHtmlElementOffset()` => `getOffset()` (add them to migration wiki)
     - [x] import SortableJS inside slick.grid instead of vanilla bundle (and other external lib, Angular-Slickgrid, ...)
     - [x] ~convert all `slick.interactions` functions to TS classes~ 
        - it brings no real benefit, except readability, and it actually caused a regression, so I'll keep the same code
     - [x] sync the newest SlickGrid CSP code to replace slick rows `innerHTML` and also slick.interactions interface changes (v5.5.3)
     - [x] replace all Event `.keyCode` with `.key` since `keyCode` was deprecated long ago,  we can use this [app](https://keyevents.netlify.app/) to find exact `key` typed or look at MDN [key values](https://developer.mozilla.org/en-US/docs/Web/API/UI_Events/Keyboard_event_key_values)
  - [x] use `grid.applyHtmlCode()` everywhere possible, instead of `innerHTML` assignment 
  - [x] review `slick.core` exports, we should only use function exports Slickgrid-Universal without releasing on downstream projects
  - [x] ~should move core into its own package~
     - won't do since GridOption interface has many references to extensions (controls/plugins) that would no longer be in the same package, we could create a core GridOption and extend it but that would overcomplicate things since we also extend them in downstream libs (Angular-Slickgrid, ...), so I think it's fine to keep it in `common/src/core`
- [x] Unit Tests coverage
  - [x] add `slick.core` full test coverage (move Utils outside)
     - [x] add test coverage for Utils moved outside of `slick.core`
  - [x] add `slick.interactions` full test coverage
  - [x] add `slick.grid` full test coverage
  - [x] add `slick.dataview` full test coverage
- [x] remove all deprecated stuff
  - [x] ~should we rename `editorOptions` and `filterOptions` to `options` for reusability or does that make no difference?~
- [x] bump TS build targets to ES2021 
- [x] cleanup Renovate, update RxJS and bump NodeJS/NPM engine requirements (we can actually drop npm engine)
- [x] rename any `commonjs` folder name to `cjs` in here and also all sub-libraries (Aurelia-Slickgrid, ...)
- [x] add pageUp/pageDown/home/end to SlickCellSelection (ref SlickGrid [PR](https://github.com/6pac/SlickGrid/pull/854) - Universal PR #1126)
   - [x] use methods from SlickGrid only available in v5, for example `hasDataView()`, `getViewportRowCount()`, `calculateAvailableSpace()`
- [x] remove any `@ts-ignore` leftover, for example I left 1 in SlickCellSelectionModel
- [x] `multiple-select-vanilla` package
  - [x] improve CSP compliance
  - [x] make an official release 1.0
    - [x] also bump build target to ES2021
    - [x] requires more work to fully support ESM (read Vite 5 [Deprecate CJS Node API](https://vitejs.dev/guide/migration.html#deprecate-cjs-node-api))
    - [x] before releasing 1.0, test it in Slickgrid-Universal
    - [x] need update on readme homepage about new CJS/ESM structure
    - [x] cannot use npm provenance because SASS copy is missing files when run in workflow vs locally (see [unpkg](https://unpkg.com/browse/multiple-select-vanilla@1.0.2-beta.4/dist/styles/sass/))
    - [x] sync with core jQuery lib that added an update to tabIndex
- [x] move `BindingEventService` into `binding` package
  - [x] ~remove `unbindByEventName` and use group name and replace by `unbindAll([group])` if possible~
  - [x] Vite 5 deprecated CJS, and might drop support for it in Vite 6. I think we need to build with either `.cjs` or `.mjs` as mentioned in Vite's [Deprecate CJS Node API](https://vitejs.dev/guide/migration.html#deprecate-cjs-node-api)
- [x] improve CSP (Content Security Policy) 
  - [x] that would require `multiple-select-vanilla` to drop the use of `innerHTML` in its code and any other inline style
  - [x] requires removing as much `innerHTML` as possible in SlickGrid/SlickDataView or use sanitizer with CSP safe approach
  - [x] add CSP Compliance documentation 
  - [x] remove any unsafe-inline style `style="..."` or `<style>...`
     - [x]  remove inline `style=""` on all Examples
  - [x] SlickDataView `compileAccumulatorLoop()` function is not CSP safe and requires `unsafe-eval` for Draggable Grouping example
- [x] using [Are the Types wrong](https://arethetypeswrong.github.io/?p=%40slickgrid-universal%2Fcommon%403.2.1) website to improve exports (ref PR #1188)
- [x] see if we can merge SlickEvent with our custom PubSubEventService and keep only 1 (possibly SlickEvent but using JS event for subscribe/notify so that we can avoid having to subscribe to all SlickGrid/DataView to later publish)
  - [x] migrate BasePubSubService to interface instead of Abstract Class
- [x] ~see if it's possible to move Plugin registering outside of grid options and keep GridOption interface with only flags options (same as core lib)~
  - [x] keep same implementation but save the resources refs earlier in the process (in the constructor), before the grid option get merged (deep copied), so this way we get to keep the singleton refs without them being recreated with different refs
- [x] add npm Provenance release workflows in CI
- [x] test the new release in Salesforce if possible
- [x] compare build size master vs next branch on the vanilla bundle zip file, this should decrease even more after removing unused SlickGrid AutoSize (since we have our own implementation in Resizer Service)
  - before: 600Kb (614,400 B)
  - after: 596Kb (610,304 B) -> after removing AutoSize 592Kb (606,208)
- [x] update readme and remove ref to 6pack/slickgrid dependency since we will drop the dependency
- [x] ~See if we should deprecate some events, i.e. we have 2x "onColumnsChanged" event GridMenu/ColumnPicker~
- [x] revisit the lib to see if there are any memory leaks (use Edge dev tool and look at snapshots) (see Angular-Slickgrid [issue](https://github.com/ghiscoding/Angular-Slickgrid/issues/625))
   - see [Debug memory leaks with the Microsoft Edge Detached Elements tool](https://blogs.windows.com/msedgedev/2021/12/09/debug-memory-leaks-detached-elements-tool-devtools/)
- [x] reimplement extendable `GridOption` and `Column` interface to avoid always having to release downstream versions every time a simple change is done in Slickgrid-Universal
   - might just need this `grid: SlickGrid<any, Column<any>, GridOption<Column<any>>> | undefined` where `Column` & `GridOption` are interfaces from downstream libs
- [x] Migrate all external libs to v4.0 (Angular-Slickgrid, Aurelia-Slickgrid, Slickgrid-React)
- [x] Aurelia-Slickgrid needs to be migrated to Aurelia 2 before migrating Slickgrid-Universal to v4.0
   - [ ] migration is done in Aurelia-Slickgrid `next` branch, but Aurelia-Slickgrid component unit tests are all missing
   - [x] possibly look at migrating to pnpm workspace for Aurelia-Slickgrid, Slickgrid-React
- [x] Changes required in external libs (`Angular-Slickgrid`, `Aurelia-Slickgrid`, `Slickgrid-React`)
  - [x] remove assignment of SortableJS to `window.Sortable` global
  - [x] SlickGrid no longer need to be in `allowedCommonJsDependencies` list on the Angular-Slickgrid lib
  - [x] ~registerPlugin inside component instead~
    - reworked this to keep same approach with grid option but keep ref in constructor before any deep copy of grid options happens (ref PR #1242)
  - [x] change minimum engine requirement to Node 18
  - [x] upgrade Angular-Slickgrid to minimum Angular 17
  - [x] implement Generic Types for Grid Options for external projects (Angular-Slickgrid, ...) should avoid breaking types when releasing 
  - [x] update Formatters Wiki after releasing new version (bunch of Formatters got removed)
  - [x] use the new PubSub approach instead of looping through SlickEvent of SlickGrid/DataView
  - [x] update Row Detail docs for Aurelia-Slickgrid that changed a lot (also update migration guide)
  - [x] add basic CSP rules in demos
  - [x] update all `registerExternalResources` to `externalResources`
  - [x] create new GitBook for each repos
    - Slickgrid-Universal GitBook admin [link](https://app.gitbook.com/o/QKZ67B9C4baxDuvuT407/s/RQhC3tvhqzyyA0mQWoBu/)
  - [x] update READMEs to drop SlickGrid dependency and add GitBook link
  - [x] update Quick Start of universal with links to other GitBook downstream libs
  - [x] update Quick Start of Aurelia-Slickgrid with Au2 approach
  - [x] add Migration link to all repos

### After the new Release
- [x] need to update Wiki after releasing new version
- [x] update all Wikis with `externalResources` instead of `registerExternalResources`
- [x] see if we can add back "open Row Detail cancelling", PR #1125 that had to be reverted because it caused a regression, need some investigation
- [x] convert Row Detail HTML strings to native HTML elements
- [x] is there a way to get RowDetail plugin working directly in Universal with a demo? That would make it easier to test the plugin instead of having to test them in external lib (Angular-Slickgrid, Aurelia-Slickgrid)
- [x] create a `version3` branch for possible last minute fixes in v3
- [x] post new release in aurelia.io discourse

### Build Size Comparison
The build sizes are compared on a Windows OS with "size on disk"
###### Slickgrid-Universal
| File                        | before | after | diff | % diff |
|---------------------|--------|-------|-----|----|
| [SF Bundle Zip File](https://github.com/ghiscoding/slickgrid-universal/tree/master/packages/vanilla-force-bundle/dist-grid-bundle-zip) | 600Kb (614,400) | 588Kb (602,112) | 12.3Kb | 2% smaller |
| [GitHub Vite Demo Website](https://ghiscoding.github.io/slickgrid-universal/) | 4.79Mb (5,029,888) | 4.78Mb (5,017,600) | 0.02Mb (or 24.6Kb) | 0.25% smaller |

**Comment by koba-ninkigumi** [comment]
Historical background aside, it is confusing that SlickGrid Universal 3 is compatible with Slickgrid Core 4 and SlickGrid Universal 4 is compatible with Slickgrid Core 5.
The next version will encompass Slickgrid Core 5. 
Would you consider this an opportunity to make it SlickGride Universal 5?

**Reply by ghiscoding** [reply]
Why would that be confusing? Technically speaking it doesn't matter what version is behind since there isn't much difference between each versions, the functionality remains the same and the only major changes were to drop jQueryUI then jQuery. User should never override the SlickGrid core version (dependency) that is pulled by Slickgrid-Universal and in the future I want to drop the SlickGrid core dependency completely and move it into Slickgrid-Universal. Having Slickgrid-Universal to no longer require SlickGrid would remove the confusion completely since they will be totally separate from each other but that is couple months away from being done

The only major differences with the next major will be
1. move SlickGrid core into Slickgrid-Universal (drop SlickGrid dependency)
2. migrating to ESM imports will indirectly drop `Slick` namespace (i.e.: `Slick.GlobalEditorLock` will have to be migrated to `SlickGlobalEditorLock` from imports and will no longer be available from global Slick window object)
